cos_alerter/daemon.py

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+
+import datetime
+import durationpy
+import json
+import signal
+import subprocess
+import sys
+import textwrap
+import time
+
+from .alerter import DataWriter, config, send_notifications
+
+
+def notify(time_data):
+    if not time.time() - time_data.notify_time > durationpy.from_str(config['notify']['repeat_interval']).total_seconds():
+        #TODO log something here
+        return
+    time_data.notify_time = time.time()
+    last_alert_time = datetime.datetime.fromtimestamp(time_data.alert_time).strftime('%A, %B %d %Y %I:%M%p %Z')
+    title = '**Alertmanager is Down!**'
+    body = textwrap.dedent(f'''
+    Your Alertmanager instance seems to be down!
+    It has not alerted COS-Alerter since {last_alert_time}.
+    ''')
+    send_notifications(title=title, body=body)
+
+
+def send_test_mail(_, __):
+    send_notifications(title='COS-Alerter test email.', body='This is a test email automatically generated by COS-alerter.')
+
+
+def sigint(_, __):
+    sys.exit()
+
+
+def main():
+
+    # Create the initial data file
+    with open(config['watch']['data_file'], 'w') as f:
+        json.dump({'alert_time': time.time(), 'notify_time': 0.0}, f)
+
+    signal.signal(signal.SIGINT, sigint)
+    signal.signal(signal.SIGUSR1, send_test_mail)
+
+    subprocess.Popen(['waitress-serve', 'cos_alerter.server:app'])
```

Comment:
Presumably we'll want this as a systemd unit in the end or something, right?

Reply:
In general, I'd say that we may as well, because turning this into a trivial deb/rpm/whatever (it won't be charmed anyway) which end-users can install on... some other system to keep an eye on thing without needing to run a container because they already have some existing infrastructure isn't hard to do, and as light as containers are, if every single application ran in one, the overhead would start to stack up.

In a "who watches the watcher who watches the watchmen" meta sense, at least if something unexpected happens, systemd will restart it. A healthcheck in the container can accomplish the same thing. [`fpm`](https://fpm.readthedocs.io/en/v1.15.1/) has good CI support, is incredibly low boilerplate for what it outputs, and is an easy way for trivial applications (like this one, in the sense that it's not compiled and has no particular dependencies from the system/kernel/udev/whatever).

For the time being, I think the question is more of "can `main()` move the logic above this into another file, or an asyncio/threaded call to spin up the server, then `main()` just runs the `while(...)` loop?"

If we did that, then using waitress inside containers can be an arg to `__main__`, and if not, maybe it's started via some other mechanism, and `main()` can also check "is the server up? If not, log a message" as part of the loop, which would go to syslog, or the container log, or whatever.

Fine as an issue, not asking for it here, but it's a little more flexible, and provides some backup for the alerter itself in case the server goes down/unresponsive or there's some other issue, where it all goes down in flames with no way to discover other than the absence.

Reply:
Created issues to track all of these.
#23
#24
#25